src/core/lib/event_engine/posix_engine/posix_endpoint.cc

```diff
@@ -549,29 +551,52 @@ void PosixEndpointImpl::MaybeMakeReadSlices() {
   }
 }
 
-void PosixEndpointImpl::HandleRead(absl::Status status) {
-  read_mu_.Lock();
+bool PosixEndpointImpl::HandleReadLocked(absl::Status status) {
   if (status.ok() && memory_owner_.is_valid()) {
     MaybeMakeReadSlices();
     if (!TcpDoRead(status)) {
       UpdateRcvLowat();
       // We've consumed the edge, request a new one.
-      read_mu_.Unlock();
-      handle_->NotifyOnRead(on_read_);
-      return;
+      return false;
     }
   } else {
-    if (!memory_owner_.is_valid()) {
-      status = absl::UnknownError("Shutting down endpoint");
+    if (!memory_owner_.is_valid() && status.ok()) {
+      status = TcpAnnotateError(absl::UnknownError("Shutting down endpoint"));
     }
     incoming_buffer_->Clear();
     last_read_buffer_.Clear();
   }
-  absl::AnyInvocable<void(absl::Status)> cb = std::move(read_cb_);
-  read_cb_ = nullptr;
-  incoming_buffer_ = nullptr;
-  read_mu_.Unlock();
-  cb(status);
+  return true;
+}
+
+void PosixEndpointImpl::HandleRead(absl::Status status) {
+  if (grpc_core::ExecCtx::Get() == nullptr) {
```

Comment:
This pattern repeats a few times. I think we should eliminate the duplication it incurs, either by locally adding extra methods, or maybe:

```
template <typename F>
void EnsureRunInExecCtx(F f) {
  if (grpc_core::ExecCtx::Get() == nullptr) {
    grpc_core::ApplicationCallbackExecCtx app_ctx;
    grpc_core::ExecCtx exec_ctx;
    f();
  } else {
    f();
  }
}
```

this could be added to exec_ctx.h and used wherever needed during this transition:

```
EnsureRunInExecCtx([]{
      grpc_core::ReleasableMutexLock lock(&read_mu_);
    if (!HandleReadLocked(status)) {
      lock.Release();
      handle_->NotifyOnRead(on_read_);
      return;
    }
    absl::AnyInvocable<void(absl::Status)> cb = std::move(read_cb_);
    read_cb_ = nullptr;
    incoming_buffer_ = nullptr;
    lock.Release();
    cb(status);
  });
```